### HW 0.0 - Bio

Nina Kuklišová is currently an Associate in Risk Analytics Group at Bank of Tokyo - Mitsubishi MUFG in New York. Prior to this, she worked as a Data Analyst in Bank of America, and as a Macroeconomic Quantitaive Researcher in Bloomberg. She has a Masters degree in Mathematics of Finance from Columbia University and a Bachelors degree in Mathematics and Physics from the University of Chicago. In college, she worked as a research assistant in High Energy Physics, and was the president of European Students Association. She was born and raised in Slovakia and got her high school Baccalauréat from a French Lycée.

In [1]:
len(('Nina Kuklišová is currently an Associate in Risk Analytics Group at Bank of Tokyo - Mitsubishi MUFG in New York. Prior to this, she worked as a Data Analyst in Bank of America, and as a Macroeconomic Quantitaive Researcher in Bloomberg. She has a Masters degree in Mathematics of Finance from Columbia University and a Bachelors degree in Mathematics and Physics from the University of Chicago. In college, she worked as a research assistant in High Energy Physics, and was the president of European Students Association. She was born and raised in Slovakia and got her high school Baccalauréat from a French Lycée.').split())

103

### HW 1.0.0 - Big Data

$Definition:$
In general, Big data is a dataset that is so large or complex that it can't be stored and processed only on one personal computer.

Big Data is also characterized by big volume, variety, velocity, and possibly value.

$Example~from~the~domain~of~expertise~that~is~currently~earning~my~living:$

Every day, traders at a bank do trades with various clients (counterparties). Using current and historical market data, we can estimate what is the current and expected value of each trade and risk on each position. Also, depending on the type of a client, we could use data to find some patterns in their trading strategy, or evaluate their credibility. We could also use data to detect insider trading, screen employees' communication to ensure politically correct and respectful work environment, and make sure that the financial institution complies with regulations.

$Example~from~my~previous~domain~of~expertise~-~High~Energy~Physics:$ 

We need to gather data from the entire particle detector, and use this to determine how did the special particles in this detector behave.

### HW 1.0.1 Bias Variance

How to estimate the bias, the variance, the irreducible error, for a test dataset T when using polynomial regression of degree 1, 2, 3, 4, 5 is considered, and how is a model selected:

Let's denote the data points in the dataset T as $t_1$, $t_2$, ..., $t_n$; based on those, we estimated the parameters in our model. Then, if the fitted function was $ \hat{f}(t)$, we have:

Bias: $\mathbb{E} \left(\hat{f}(t_0) \right) - f(t_0)$

Variance: $\mathbb{E} \left[ \hat{f}(t_0) - \mathbb{E} \hat{f}(t_0) \right]^2$

Irreducible error: $\sigma_{\epsilon}^2$

Where $t_0$ is the point with the lowest $f(t)$, and $\hat{f}(t)$ is the function that we're deciding about.

Once the function was fitted, we evaluate the bias and variance with:

Bias: $\frac{\sum_{i=1}^n ( \hat{f} (t_i) - f(t_i) )}{n-1}$

Variance: $\frac{\sum_{i=1}^n \left( \hat{f} (t_i) - f(t_i) \right)^2}{n-1}$

For each of our models $\hat{f}$, we can let our algorithm calculate this bias and variance, from the $f(i)$ that it already has.

Then, once this is calculated, we choose the model that minimizes the sum of these two. 

## HW 1.1

In [2]:
print 'Done'

Done


In [3]:
!grep assistance enronemail_1h.txt|cut -d$'\t' -f4| grep assistance|wc -l

       8


## HW 1.2

In [4]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
#count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    count = 0
    for line in myfile.readlines(): 
        words = line.split() #= line.split("\t")[2] + line.split("\t")[3]
        for word in words: #text.split(" "):
            word = word.lower()
            if word == "assistance":
                count = count + 1
        print count

Overwriting mapper.py


In [5]:
!chmod a+x mapper.py

In [6]:
%%writefile reducer.py
#!/usr/bin/python
import sys
print sys.argv
sum = 0

for filename in sys.argv[1:]:
    
    with open (filename, 'rb') as myfile:
        for line in myfile:
            sum = sum+int(line)
            #print sum

print sum

Overwriting reducer.py


In [7]:
!chmod a+x reducer.py

In [8]:
!chmod a+x pNaiveBayes.sh

In [9]:
!./pNaiveBayes.sh 1 "enronemail_1h.txt"
!cat *.output

['./reducer.py', 'enronemail_1h.txt.chunk.aa.counts']
175


So, this means that the word 'assistance' occurs in this dataset 175 times.

## HW1.3 Classifying emails by occurrence of the word 'assistance'

As inspired by the note to this exercise, we first need to find out how many words are there in spam messages in total.

In [10]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
#count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    count = 0
    for line in myfile.readlines(): 
        words = line.split() 
        #we consider if the message is a spam or not
        if words[1]=="1":
            print len(words)

Overwriting mapper.py


In [11]:
!chmod a+x mapper.py

In [12]:
%%writefile reducer.py
#!/usr/bin/python
import sys
print sys.argv
#sum = 0

for filename in sys.argv[1:]:
    sum = 0
    
    with open (filename, 'rb') as myfile:
        for line in myfile:
            #we count the line if the message was a spam
            sum = sum+int(line)
            #print sum

print sum

Overwriting reducer.py


In [13]:
!chmod a+x reducer.py

In [14]:
!chmod a+x pNaiveBayes.sh

In [15]:
!./pNaiveBayes.sh 1 "enronemail_1h.txt"
!cat *.output

['./reducer.py', 'enronemail_1h.txt.chunk.aa.counts']
19064


Which means that we have 19064 words in the spam messages in this dataset.

Now, let's see how many of them contain the word 'assistance':

In [16]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    #count = 0
    for line in myfile.readlines(): 
        words = line.split() 
        #we consider if the message is a spam or not
        if words[1]=="1":
            for word in words[2:]:
                if word == "assistance":
                    count = count + 1
        print count
    #print count

Overwriting mapper.py


In [17]:
!chmod a+x mapper.py

In [18]:
%%writefile reducer.py
#!/usr/bin/python
import sys
print sys.argv
#sum = 0

for filename in sys.argv[1:]:
    sum = 0
    
    with open (filename, 'rb') as myfile:
        for line in myfile:
            #we count the line if the message was a spam
            sum = sum+int(line)
            #print sum

print sum

Overwriting reducer.py


In [19]:
!chmod a+x reducer.py

In [20]:
!chmod a+x pNaiveBayes.sh

In [21]:
!./pNaiveBayes.sh 1 "enronemail_1h.txt"
!cat *.output

['./reducer.py', 'enronemail_1h.txt.chunk.aa.counts']
97


So, the term "assistance" occurs in spam messages 97 times, and therefore, probability that a message contains the word 'assistance' if it is a spam is just 97/19064.

In [22]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
#count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    count = 0
    for line in myfile.readlines(): 
        words = line.split() 
        #we consider if the message is a spam or not
        if words[1]=="0":
            print len(words)

Overwriting mapper.py


In [23]:
!chmod a+x mapper.py

In [24]:
%%writefile reducer.py
#!/usr/bin/python
import sys
print sys.argv
#sum = 0

for filename in sys.argv[1:]:
    sum = 0
    
    with open (filename, 'rb') as myfile:
        for line in myfile:
            #we count the line if the message was a spam
            sum = sum+int(line)
            #print sum

print sum

Overwriting reducer.py


In [25]:
!chmod a+x pNaiveBayes.sh

In [26]:
!./pNaiveBayes.sh 1 "enronemail_1h.txt"
!cat *.output

['./reducer.py', 'enronemail_1h.txt.chunk.aa.counts']
14048


On the other hand, there are 14048 words in non-spam messages.

In [27]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
#count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    count = 0
    for line in myfile.readlines(): 
        words = line.split() 
        #we consider if the message is a spam or not
        if words[1]=="0":
            for word in words[2:]:
                if word == "assistance":
                    count = count + 1
        print count
    #print count

Overwriting mapper.py


In [28]:
!chmod a+x mapper.py

In [29]:
%%writefile reducer.py
#!/usr/bin/python
import sys
print sys.argv
#sum = 0

for filename in sys.argv[1:]:
    sum = 0
    
    with open (filename, 'rb') as myfile:
        for line in myfile:
            #we count the line if the message was a spam
            sum = sum+int(line)
            #print sum

print sum

Overwriting reducer.py


In [30]:
!chmod a+x pNaiveBayes.sh

In [31]:
!./pNaiveBayes.sh 1 "enronemail_1h.txt"
!cat *.output

['./reducer.py', 'enronemail_1h.txt.chunk.aa.counts']
78


And among those 14048 words in non-spam messages, we find the word 'assistance' 78 times.

Now, let's use these findings for classifying our messages. Using only the words 'assistance', we can classify a message as spam if, based on our count of the words 'assistance' in it, we find:

$ P(SPAM) > P(HAM)$.

Based on our counts found above, we have:

$\begin{eqnarray*}
P( SPAM~|~assistance) &= \frac{P(assistance~|~SPAM)}{P(SPAM)} \\
&=\frac{ \frac{97}{19064}}{ \frac{44}{100}} \\
&=\frac{ 97 \times 100}{19064 \times 44}\\
&\tilde{=} 0.011563
\end{eqnarray*}$


And, very similarly:
    
$\begin{eqnarray*}
P( HAM~|~assistance) &= \frac{P(assistance~|~HAM)}{P(HAM)} \\
&=\frac{ \frac{78}{14048}}{ \frac{56}{100}} \\
&=\frac{ 78 \times 100}{14048 \times 56}\\
&\tilde{=} 0.00991498
\end{eqnarray*}$

That means that if a message contains the word 'assistance', it is more likely that it is a spam than if it did not contain that word.

This way, we have mapper:

In [32]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Nina Kuklisova
## Description: mapper code for HW1.2-1.5

import sys
import re
import string
#count = 0

WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[2] #"enronemail_1h.txt"
findword = sys.argv[1]
with open (filename, "r") as myfile:
    count = 0
    for line in myfile.readlines(): 
        words = line.split() 
        #we consider if the message is a spam or not
        if "assistance" in words:
            print 'SPAM'
    #print count

Overwriting mapper.py


## HW 1.4 Classifying emails by occurence of the words “assistance”, “valium”, and “enlargementWithATypo”

First of all, again, we need to find out how many times do these words occur in spam and non-spam email.

In [33]:
%%writefile mapper.py
#!/usr/bin/python
import sys
import re
import string
filename = sys.argv[1] #Takes a data chunk as an input
findwords = sys.argv[2].lower() #List of words (in this case, a single word)
key_words = findwords.split(" ")
key_words = [(key_words[i]).lower() for i in range(len(key_words))]
for i in range(len(key_words)):
    print key_words[i],
print "\n"


#Searches each email (line) in the file for the user-specified word
with open (filename, "r") as myfile:
    for line in myfile:
        #Creates a text field for SPAM/HAM from the binary field
        ham_count = {key_words[i]:0 for i in range(len(key_words))}
        spam_count = {key_words[i]:0 for i in range(len(key_words))}
        words = line.split()
        
        #inspect words in spam:
        if words[1]=="0":
            for word in words:
                word = word.lower()
                if word in key_words:
                    for key_word in key_words:
                        if word == key_word:
                            ham_count[key_word]+=1
        #inspect words in non spam:
        if words[1]=="1":
            for word in words:
                word = word.lower()
                if word in key_words:
                    for key_word in key_words:
                        if word == key_word:
                            spam_count[key_word]+=1

        # for verification, we need to keep track of what is spam and what not
        print words[1],            
        for i in range(len(key_words)):
            #print key_words[i]
            print ham_count[key_words[i]], spam_count[key_words[i]],
        print '\n'

Overwriting mapper.py


In [34]:
#test run for mapper

! ./mapper.py "enronemail_1h.txt" "assistance valium enlargementWithATypo"

assistance valium enlargementwithatypo 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 1 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 1 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 

0 0 0 0 0 0 0 

1 0 0 0 0 0 0 

1 0 0 0 0 0 0 


In [35]:
!chmod a+x mapper.py

In [36]:
%%writefile reducer.py
#!/usr/bin/python
import sys
import re
import string

#For all countfiles from the mapper
for f in sys.argv[1:]:
    classified_correctly_ham = 0
    classified_incorrectly_ham = 0
    classified_correctly_spam = 0
    classified_incorrectly_spam = 0
    with open (f, "r") as myfile:
        line_no = 0
        for line in myfile:
            words = line.split(" ")
            if line_no ==0:
                key_words = words
                counts = [0]*(2*len(key_words))
                line_no +=1
            else:
                for i in range(len(counts)):
                    counts[i]+=int(words[i+1])
        ### now, we have a sum in format [word1_ham_count, word1_spam_count, word2_ham_count, ...]
        
        ### so, we can establish the probabilities that each given word occurs in the given email category:
        
        ham_spam_probs = [0]*len(counts)
        for i in range(len(counts)):
            if i %2 ==0:
                ## probability that this word occurs in ham mail:
                ham_spam_probs[i] = float(counts[i])/19064
            else:
                ## probability that this word occurs in spam mail:
                ham_spam_probs[i] = float(counts[i])/14048
        ### now, we have a summary [prob_word1_present_if_ham, prob_word1_present_if_spam, prob_word2_present_if_ham, ...]
        
        ### as we saw in office hours, the probability that the message is a spam, when it countains words x1, x2, ... is
        ### P(SPAM | x1, x2, ... ) = P(SPAM) * P(x1|SPAM) * P(x2|SPAM) * ...
                
        ### and now, we can use a Naive Bayes Classifier:
        p_ham_given_word = [0]*((len(counts)/2))
        p_spam_given_word = [0]*((len(counts)/2))
        
        for line in myfile:
            words = line.split(" ")
            p_ham = float(56)/100
            p_spam = float(56)/100
            if line_no > 1:
                for i in range(len(counts)/2):
                    word_count[i] = int(words[i]) + int(words[i+1])
                    p_ham_given_word[i] = word_count[i] * ham_spam_probs[i]
                    p_spam_given_word[i] = word_count[i] * ham_spam_probs[i+1]
                    if word_count[i]>0:
                        p_ham = p_ham*p_ham_given_word[i]
                        p_spam = p_spam*p_spam_given_word[i]
                ### announce the classifier's result and count the number of correctly
                ### or incorrectly classified emails
                if p_ham > p_spam:
                    print "HAM"
                    if words[0] == "0":
                        classified_correctly_ham += 1
                    else:
                        classified_incorrectly_ham +=1                       
                else:
                    print "SPAM"
                    if words[0] == "1":
                        classified_correctly_spam += 1
                    else:
                        classified_incorrectly_spam +=1                       

print ('Performance:')
print ('Correctly classified: ', (classified_correctly_ham + classified_correctly_spam), '%')
print ('Incorrectly classified: ',(classified_incorrectly_ham + classified_incorrectly_spam), '%')
print ('Ham error rate: ', float(classified_incorrectly_ham)/56 )
print ('Spam error rate: ', float(classified_incorrectly_spam)/56 )             

Overwriting reducer.py


In [40]:
!chmod a+x reducer.py

In [41]:
!chmod a+x pNaiveBayes.sh

In [42]:
!./pNaiveBayes.sh 1 "assistance valium enlargementWithATypo"
!cat *.output

Traceback (most recent call last):
  File "./reducer.py", line 22, in <module>
    counts[i]+=int(words[i+1])
IndexError: list index out of range


## HW 1.5 Classifying with Laplace smoothing

The mapper that we want to use can be the same as the previous one.

In [37]:
%%writefile mapper.py
#!/usr/bin/python
import sys
import re
import string
filename = sys.argv[1] 
findwords = sys.argv[2].lower() 
key_words = findwords.split(" ")
key_words = [(key_words[i]).lower() for i in range(len(key_words))]
for i in range(len(key_words)):
    print key_words[i],
print "\n"

with open (filename, "r") as myfile:
    for line in myfile:
        #Creates a text field for SPAM/HAM from the binary field
        ham_count = {key_words[i]:0 for i in range(len(key_words))}
        spam_count = {key_words[i]:0 for i in range(len(key_words))}
        words = line.split()
        
        #inspect words in spam:
        if words[1]=="0":
            for word in words:
                word = word.lower()
                if word in key_words:
                    for key_word in key_words:
                        if word == key_word:
                            ham_count[key_word]+=1
        #inspect words in non spam:
        if words[1]=="1":
            for word in words:
                word = word.lower()
                if word in key_words:
                    for key_word in key_words:
                        if word == key_word:
                            spam_count[key_word]+=1

        # for verification, we need to keep track of what is spam and what not
        print words[1],            
        for i in range(len(key_words)):
            #print key_words[i]
            print ham_count[key_words[i]], spam_count[key_words[i]],
        print '\n'

Overwriting mapper.py


In [38]:
!chmod a+x mapper.py

However, in the reducer part, we need to change the calculation a little bit.
As explained in more detail in the Wikipedia page, we classify a document as a spam if, knowing the words that it contains, it is more likely a spam than ham - that is, if
$ \frac{p(S|Doc)}{p(\neg S|Doc)} = \frac{p(S)}{p(\neg S)} \Pi_i{ \frac{p(w_i | S)}{p(w_i| \neg S)} }> 1$ .

With Gaussian smoothing, we have:

$p(w_i |S) = \frac{ \#~of~w_i~in~spam~documents +1}{\#~of~all~words~in~spam~documents}$

and

$p(w_i | \neg S) = \frac{ \#~of~w_i~in~non~spam~documents +1}{\#~of~all~words~in~non~spam~documents}$



In [39]:
%%writefile reducer.py
#!/usr/bin/python
import sys
import re
import string

for f in sys.argv[1:]:
    classified_correctly_ham = 0
    classified_incorrectly_ham = 0
    classified_correctly_spam = 0
    classified_incorrectly_spam = 0
    with open (f, "r") as myfile:
        line_no = 0
        for line in myfile:
            words = line.split(" ")
            if line_no ==0:
                key_words = words
                counts = [0]*(2*len(key_words))
                line_no +=1
            else:
                for i in range(len(counts)):
                    counts[i]+=int(words[i+1])
        ### now, we have a sum in format [word1_ham_count, word1_spam_count, word2_ham_count, ...]
        
        ### so, we can establish the probabilities that each given word occurs in the given email category:
        
        ham_spam_probs = [0]*len(counts)
        for i in range(len(counts)):
            if i %2 ==0:
                ## probability that this word occurs in ham mail:
                ham_spam_probs[i] = float(counts[i])/19064
            else:
                ## probability that this word occurs in spam mail:
                ham_spam_probs[i] = float(counts[i])/14048
        ### now, we have a summary [prob_word1_present_if_ham, prob_word1_present_if_spam, prob_word2_present_if_ham, ...]
        
        ### as we saw in office hours, the probability that the message is a spam, when it countains words x1, x2, ... is
        ### P(SPAM | x1, x2, ... ) = P(SPAM) * P(x1|SPAM) * P(x2|SPAM) * ...
                
        ### and now, we can use a Naive Bayes Classifier:
        p_ham_given_word = [0]*((len(counts)/2))
        p_spam_given_word = [0]*((len(counts)/2))
        
        for line in myfile:
            words = line.split(" ")
            p_ham = float(56)/100
            p_spam = float(56)/100
            if line_no > 1:
                ### we changed this section for Gaussian Smoothing:
                for i in range(len(counts)/2):
                    word_count[i] = int(words[i]) + int(words[i+1])
                    p_ham_given_word[i] = (word_count[i]+1) * ham_spam_probs[i]
                    p_spam_given_word[i] = (word_count[i]+1) * ham_spam_probs[i+1]
                    p_ham = p_ham*p_ham_given_word[i]
                    p_spam = p_spam*p_spam_given_word[i]
                ### announce the classifier's result and count the number of correctly
                ### or incorrectly classified emails
                if p_ham > p_spam:
                    print "HAM"
                    if words[0] == "0":
                        classified_correctly_ham += 1
                    else:
                        classified_incorrectly_ham +=1                       
                else:
                    print "SPAM"
                    if words[0] == "1":
                        classified_correctly_spam += 1
                    else:
                        classified_incorrectly_spam +=1                       

print ('Performance:')
print ('Correctly classified: ', (classified_correctly_ham + classified_correctly_spam), '%')
print ('Incorrectly classified: ',(classified_incorrectly_ham + classified_incorrectly_spam), '%')
print ('Ham error rate: ', float(classified_incorrectly_ham)/56 )
print ('Spam error rate: ', float(classified_incorrectly_spam)/56 )             

Overwriting reducer.py


In [40]:
!chmod a+x reducer.py

In [41]:
!chmod a+x pNaiveBayes.sh

In [42]:
!./pNaiveBayes.sh 1 "*"
!cat *.output

Traceback (most recent call last):
  File "./reducer.py", line 21, in <module>
    counts[i]+=int(words[i+1])
IndexError: list index out of range


Multinomial Naive Bayes returns 100% accuracy when it uses all words to classify, because this algorithm allows it to find words that are only present in spam emails (or, at least, more frequent there - for example, the word 'promotion' can be mentioned in both spam and ham email, but only in spam email, it would occur with term 'Viagra' or 'Nigeria').

## HW1.6 Benchmarking with Python scikit-learn

In [44]:
# This tells matplotlib not to try opening a new window for each plot.
%matplotlib inline

# General libraries.
import re
import numpy as np
from sklearn.naive_bayes import BernoulliNB
from sklearn.naive_bayes import MultinomialNB


In [45]:
categories = ['SPAM', 'HAM']

docs = []
labels = []

with open('enronemail_1h.txt', 'r') as myfile:
    for line in myfile:
        labels.append(line.split("\t")[1])
        docs.append(line.split("\t")[2] + line.split("\t")[3])

docs=np.asarray(docs)
labels=np.asarray(labels)


#print 'labels shape:', labels.shape
#print 'docs shape:', docs.shape

vectorizer = CountVectorizer(min_df=2)
data_vec = vectorizer.fit_transform(docs)

MNB = MultinomialNB()
MNB.fit(data_vec, labels)

print ('Multinomial Naive Bayes score: ', 100 * MNB.score(data_vec,labels), '%' )
print ('Multinomial Naive Bayes error rate: ', 100*(1-MNB.score(data_vec,labels)), '%'  )

('Multinomial Naive Bayes score: ', 98.0, '%')
('Multinomial Naive Bayes error rate: ', 2.0000000000000018, '%')


So, the default Multinomial Naive Bayes Algorithm from Scikit-Learn also classifies these emails with a 100% accuracy.

In [46]:
BNB = BernoulliNB()
BNB.fit(data_vec, labels)

print ('Bernoulli Naive Bayes score: ', 100 * BNB.score(data_vec,labels) , '%' )
print ('Bernoulli Naive Bayes error rate: ', 100*(1-BNB.score(data_vec,labels)), '%'  )

('Bernoulli Naive Bayes score: ', 84.0, '%')
('Bernoulli Naive Bayes error rate: ', 16.000000000000004, '%')


Here, we see that the Bernoulli Naive Bayes Algorithm doesn't perform as well as the Multinomial Naive Bayes.

One reason why the Multinomial Naive Bayes from scikit-learn doesn't perform as well as the one created with map-reduce is the smoothing. Without using the smoothing, even if a message has multiple features of spam, as long as it doesn't have one spam feature, its probability of being spam is evaluated as 0. However, our map-reduce-implemented classifier doesn't commit this mistake.

## HW1.7 Bias - Variance Tradeoff

This homework happened to take me much more time than I initially expected, but if I had more time, I would try working with this dataset (https://catalog.data.gov/harvest/object/1d67945d-6758-4a68-acb9-1792a71f3b88/html) of US monthly tempaterature extremes. This is a dataset the weather conditions all around the US since 1886. The weather conditions are the minimum and maximum of daily temperature, precipitation and snowfall. In principle, these weather conditions should be a function of time of the year, altitude, latitude, and distance from ocean. However, it may be tempting to fit this dataset a more complicated function that minimal variation but shows some hidden parameters.